src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHash.Apple.cs

```diff
@@ -0,0 +1,163 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Security.Cryptography.Apple;
+
+using PAL_HashAlgorithm = Interop.AppleCrypto.PAL_HashAlgorithm;
+
+namespace System.Security.Cryptography
+{
+    internal static partial class LiteHashProvider
+    {
+        private static LiteHash CreateHash(string hashAlgorithmId)
+        {
+            PAL_HashAlgorithm algorithm = HashAlgorithmNames.HashAlgorithmToPal(hashAlgorithmId);
+            return new LiteHash(algorithm);
+        }
+
+        private static LiteHmac CreateHmac(string hashAlgorithmId, ReadOnlySpan<byte> key)
+        {
+            PAL_HashAlgorithm algorithm = HashAlgorithmNames.HashAlgorithmToPal(hashAlgorithmId);
+            return new LiteHmac(algorithm, key);
+        }
+    }
+
+    internal readonly struct LiteHash : ILiteHash
```

Comment:
LiteHash looks a lot like HashProvider.  Why do we need both?

Reply:
* LiteHmac doesn't keep a key copy for resets
* LiteHash / LiteHmac don't reset once used.
* LiteHash / LiteHmac don't keep track of whether operations are in-flight with `_running`
* It's a struct

_Could we put a `Reset` on these (taking the key for the HMAC variant) and have `HashProvider` wrap those?_

Maybe. That's kind of what I did for the symmetric one-shots (replacing IV for HMAC keys). I avoided that here because I felt that the structs were simple enough on their own, really just a few methods on a `SafeHandle`. De-duplicating it would just mean another level of indirection on the HashProviders to the Lite ones.

Reply:
Hm.  Feels like the HashProvider could just have a LiteHash as a field, and then manage the state on top of that.

If it doesn't nest easily, then it's not a whole lot of duplication... but it's still nice to only need to make changes N times (Windows, Unix, Android, Apple OSes, Browser) instead of N * M (when M > 1)... so I have to ask :smile:

Reply:
This "works" for Unix / Android and Apple. It works less-so to me for Windows. The Windows one was already somewhat specialized, and it does upfront work to determine if the platform supports usable CNG hashes.

What didn't work either was a single `HashProvider` for multiple platforms, even if they wrapped their own `LiteHash`, since everyone wants their own state.

If the last two commits are just seem like noise without pushing far enough in to the "make it less duplicated", I can revert or drop them.

The alternative, if we really do not like the duplication is to make the stream one shots allocate a bunch more and abandon the lite hashes altogether. Though I believe the one-shots should remain as light as reasonably possible, even if they do end up allocating a safe-handle or two.

Reply:
While "no duplication" would be ideal, "less duplication" is at least virtuous.  Thanks for pursuing it.